pkg/tenant/ensign.go

```diff
@@ -0,0 +1,201 @@
+package tenant
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rotationalio/ensign/pkg/tenant/db"
+	"github.com/rotationalio/ensign/pkg/utils/metatopic"
+	sdk "github.com/rotationalio/go-ensign"
+	api "github.com/rotationalio/go-ensign/api/v1beta1"
+	mimetype "github.com/rotationalio/go-ensign/mimetype/v1beta1"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+// TopicSubscriber is a struct with a go routine that subscribes to the Ensign
+// "meta topic" topic and applies asynchronous updates to the topics in the Tenant
+// database.
+type TopicSubscriber struct {
+	client *EnsignClient
+	stop   chan struct{}
+}
+
+func NewTopicSubscriber(client *EnsignClient) *TopicSubscriber {
+	return &TopicSubscriber{
+		client: client,
+	}
+}
+
+// Run the topic subscriber under the waitgroup.
+func (s *TopicSubscriber) Run(wg *sync.WaitGroup) error {
+	if s.stop != nil {
+		return errors.New("topic subscriber is already running")
+	}
+
+	if wg != nil {
+		wg.Add(1)
+	}
+
+	s.stop = make(chan struct{})
+	go func() {
+		s.Subscribe()
+		s.stop = nil
+		if wg != nil {
+			wg.Done()
+		}
+	}()
+	return nil
+}
+
+// Stop the topic subscriber.
+func (s *TopicSubscriber) Stop() {
+	if s.stop != nil {
+		close(s.stop)
+	}
+}
+
+func (s *TopicSubscriber) Subscribe() {
+	var (
+		err error
+		sub *sdk.Subscription
+	)
+
+	// Wait until the Ensign server is ready
+	if attempts, err := s.client.WaitForReady(); err != nil {
+		// Note: Using WithLevel with FatalLevel does not exit the program but this is
+		// likely a critical configuration error that we want to fix immediately.
+		log.WithLevel(zerolog.FatalLevel).Int("attempts", attempts).Err(err).Msg("could not connect to ensign server")
+		return
+	}
```

Comment:
So this does not crash the tenant process; just logs a critical message but tenant continues to operate right? 

I guess, I'm wondering where we should put this wait for ready -- I think it should block the tenant process serving; right now this just heads off on its own go routine during startup … and tenant starts serving but it can be up to ten minutes (when many topics can be created and updated) before we're alerted that it's not connected to Ensign. 

the paradigm I've been using is to have tenant stall and not report to k8s that it's ready until it can connect to tenant; I propose we merge this PR in and rearchitect the readiness to follow that same pattern for connecting to Ensign.

Reply:
That makes more sense to me. I'll fix the waitgroup issue and then merge this in.